database.cc

```diff
@@ -1797,6 +1797,10 @@ database::stop() {
         return _streaming_dirty_memory_manager.shutdown();
     }).then([this] {
         return _memtable_controller.shutdown();
+    }).then([this] {
+        return _user_sstables_manager->close();
+    }).then([this] {
+        return _system_sstables_manager->close();
```

Comment:
The is a destructor-like sequence, failures are not recoverable. I can change to a finally() to reduce worries about this.

Reply:
I changed my mind. A finally() just casts doubt on the operations it protects, and makes the code using it look questionable if it doesn't handle or at least log failures. So it creates a lot of bureaucracy, for something that is a bug (leaking an exception via close()).

Reply:
We have some close() functions which deliberately can throw. For example, file::close() if I remember correctly is designed such that it can throw - the file will *in any case* be closed by close(), but additionally it can throw if there was some error in the process. Maybe this was always a mistake :-( I'll give this some more thought and perhaps open a separate issue about that.  In any case I'll resolve this discussion.

Reply:
file::close() mirrors close(2). I suppose it was a mistake, since correct usage should flush the file right before closing to collect any error, then close.

Note that while close(2) can return an error, it is not guaranteed to, since some operations can still happen in the background (and there can be other handles for that file description), so we can definitely mandate a flush() before closing any file that was written to.

Reply:
@avikivity I added at the bottom of https://github.com/scylladb/seastar/issues/724 a discussion on what does failure in `close()` mean. My conclusion is that if a big object holds multiple independent objects, it should `close()` all of them. In this case if we don't call `_system_sstables_manager.close()`, because the previous one failed, it is not safe to destroy "this". So I do think this code is unsafe. So not resolving this discussion just yet...

Reply:
But what does it mean that _user_sstables_manager.close() failed? We're already broken.

Reply:
I won't insist here, because I want to commit your series and I don't think this problem will ever happen, and because the problem is pre-existing, but in theory it is possible that the first close() will fail because of some IO error in flushing one type of sstables, and it doesn't mean we shouldn't even try to flush some other types of sstable. Or something like that.

Reply:
No, those failures are logged and converted to successes (since they happen long after the actions that initiated them).

Note those closes happen for read-only sstables, or for writing sstables, but only after a flush was performed in the foreground.